### PR TITLE
Changes Space Ninja dash into rewritten Syndicate Experimental Teleporter dash

### PR DESCRIPTION
**Pull request by aramix273**
## About The Pull Request

**NEW!** summary of changes:
- logic in syndicate teleporter was extracted into its own ``do_dash`` proc
- this ``do_dash`` proc was used to rewrite the syndicate teleporter (no functional change), and...
- ... also the ninja dash, which is no longer a teleport, instead:
  - dashes apply structure damage to all structures in path (not walls though)
  - if it fails to break obstacles, the dash is stopped early
  - active bluespace anchors will also be able to stop these dashes inside or into their area of effect
  - additionally, any mob encountered in the path will be automatically attacked with the katana, unless on help intent
- to accomodate the last combat-oriented change, katana damage was nerfed from 40 to 25
- dash charges used to happen in parallel, meaning you could use it like 1,2,3-wait for one recharge-1,2,3 again; now they recharge sequentially, serial, one at a time (see PR comments for videos showing the difference)
- ninja suit examine no longer infinitely loops (recursion even) and now properly shows some stats when initialized
- other runtime fixes related to loss of functionality or visual effects

intention of PR is to balance ninja's dash - main change is replacing the teleport with a jaunt similar to syndicate experimental teleporter's bluespace wake: it now applies 150 damage to structures and stops the dash if the damage does not break them; the telefrag effect is replaced with a suggested auto-slash, which does not hit if the ninja is on help intent

to avoid code repetition, the experimental teleporter's "dash" code was extracted into a generalized, reusable proc called ``do_dash``, which takes a movable atom, source and destination turfs, goes through each of them in a straight line and applies structure damage, then calls a separate callback for each turf crossed
it has several parameters for object damage, the callback, teleport channel and whether non-broken structures stop the jaunt or not
with this, the experimental teleporter's code was rewritten to have equivalent function and is as a result cleaner
besides this use example, the proc contains comments and a slight doc for how it's intended to be used

with this ``do_dash`` proc, ninja's dash was replaced with settings similar to the experimental teleporter's and given a callback to automatically attack all mobs crossed through with the sword, or only taps them if it's on help intent
because of the default teleport channel used, bluespace anchors now also block the dash up until its range

another small change is fixing one character in ninja suit examine proc that made it infinitely loop and fail

TODOs:
- [ ] consider a better chat message for those crossed by ninja dash
- [x] discuss balancing in parallel with #11583
- [ ] check that bluespace anchor stops dash at the limit of its jaunt 

## Why It's Good For The Game

- balances ninja's teleports to make them less spammable and offer opportunities to chase (mesons will become less viable as wall phasing is removed)
- offers autoattack on dash at cost of nerfed energy katana damage
- the code is more clean as a result of the changes and the new proc can be reused in other features
- fixes several related runtimes leading to loss of functionality or of visual effects

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

ninja suit examine
image

experimental syndicate teleporter

https://github.com/user-attachments/assets/a3457c37-8899-4d00-8953-88f2a5703d2a

edge cases such as teleporting at edge of non-looping zlevel or on teleport-restricted zlevel also tested

image

ninja dash testing

https://github.com/user-attachments/assets/743bdd86-14a7-4dea-ac65-78da273428d2

ninja dash slash

https://github.com/user-attachments/assets/a1bd832e-eeeb-4775-8381-dd1663f01a2d

ninja dash testing 2

https://github.com/user-attachments/assets/146ff4ce-43e6-4394-9bea-f6ff5b5b1cd1

</details>

## Changelog
:cl: Aramix
balance: replaced ninja's teleport with a dash similar to syndicate experimental teleporter's, but instead of telefrag, it automatically attacks every mob encountered with the sword, unless on help intent
balance: energy katana damage nerfed from 40 to 25
balance: ninja dash charges no longer recharge in parallel, now recharging one at a time
balance: ninja dash can now be stopped by a functional bluespace anchor
fix: fixed runtimes causing loss of functionality or of visual effects with energy katana dash
fix: ninja suit examine now properly shows related stats when initialized
refactor: extracted the logic in syndicate experimental teleporter's attempt_teleport into a generalized and parametrizable do_dash proc
code: rewrote syndicate experimental teleporter and ninja dash to use the new do_dash proc
/:cl:

### CODE REVIEW

**Comment by PowerfulBacon** [comment]
With do_item_attack_animation, you can add the item attack animation for mobs that you pass through.

**Comment by Rukofamicom** [comment]
This is such a significant nerf to the ninja that headmins should probably be notified to consider changing their standing from "Self-Defense only" to "Murderbone" in the rules.

**Comment by llol111** [comment]
The primary appeal of ninja, to me, was always the supreme mobility and not their killing potential.
Having super kill powers with the huge katana damage is also great fun, but that wasnt what ninjas *soul* is to me.
So this is a bit sad tbh, not a fan

Ultimately not *that* bad though, no more hopping over wall will mean thatthe role will become somewhat more combat focused imo, but aslong as they're still badass at that atleast thats fineish

**Comment by Fronsis** [comment]
> The primary appeal of ninja, to me, was always the supreme mobility and not their killing potential. Having super kill powers with the huge katana damage is also great fun, but that wasnt what ninjas _soul_ is to me. So this is a bit sad tbh, not a fan
> 
> Ultimately not _that_ bad though, no more hopping over wall will mean thatthe role will become somewhat more combat focused imo, but aslong as they're still badass at that atleast thats fineish

As someone who has discussed this topic multiple times, alongside being a Security main that has experienced first hand with other seccies one of the main points of classifying the ninja being unfun to play against were:

- Constant Dash free out of jail cards
- General damage and kit, etcetera won't elaborate to not make it long

Ultimately if this gets merged alongside #11583 it will be more balanced and will make the Ninja player to think where and when to fight instead of being able to disengage at any given time with no punishment.

I agree though that i'd rather have mobility and stealth than a 2 crit katana slash gamer.

**Comment by SunShine214** [comment]
Due to how heavy this nerf is, ninjas will need ways to heal/stop blood loss easier. Currently the only way to do so is to use piercing syringes (pretty metagamey) or take of the suit with the minute long timer it has to deactivate it. Without the ability to run away well, ninjas will lack the time to heal. Additionally, I would not like to see ninja become another antag sec can defeat in less then 5 minutes with roundstart gear, as almost all our antags are part of that catagory.

**Comment by Fronsis** [comment]
> Due to how heavy this nerf is, ninjas will need ways to heal/stop blood loss easier. Currently the only way to do so is to use piercing syringes (pretty metagamey) or take of the suit with the minute long timer it has to deactivate it. Without the ability to run away well, ninjas will lack the time to heal. Additionally, I would not like to see ninja become another antag sec can defeat in less then 5 minutes with roundstart gear, as almost all our antags are part of that catagory.

Current ninja falls on the category where they can kill literally the most armed security staff due to their kit, there's no in between and it's been untouched for a long time.

If your concern is healing then on one part it's more of a player issue, they need to pick their fights wisely instead of relying on their absurd damage to demolish any threats and then retreat to heal, on the other hand a lot of people mentioned the bleeding issue while discussing #11583 on discord and suggested that it should be at least possible to bandage yourself.

**Comment by SunShine214** [comment]
> > Due to how heavy this nerf is, ninjas will need ways to heal/stop blood loss easier. Currently the only way to do so is to use piercing syringes (pretty metagamey) or take of the suit with the minute long timer it has to deactivate it. Without the ability to run away well, ninjas will lack the time to heal. Additionally, I would not like to see ninja become another antag sec can defeat in less then 5 minutes with roundstart gear, as almost all our antags are part of that catagory.
> 
> Current ninja falls on the category where they can kill literally the most armed security staff due to their kit, there's no in between and it's been untouched for a long time.
> 
> If your concern is healing then on one part it's more of a player issue, they need to pick their fights wisely instead of relying on their absurd damage to demolish any threats and then retreat to heal, on the other hand a lot of people mentioned the bleeding issue while discussing #11583 on discord and suggested that it should be at least possible to bandage yourself.

Primarily I am concerned there will longer be as much tolerance for failure; ninja is an antag you could have fun in and gimmick while being terrible at antag/unskilled at combat. Of course, this also means skilled players become an issue, but I hope there is a midground that can be achievable.

**Comment by PowerfulBacon** [comment]
I would much rather merge this PR in isolation to the other PR and then see how ninja plays rather than merge the alternate balancing PR which just touches everything at once for the sake of touching it without really realising what makes the ninja powerful.

My request for thus PR is:
- Lower the katana damage to ~20-25
- Fix the recharging issues

Then I will be happy to merge it and see how it plays out.

**Comment by aramix273** [comment]
several aspects discussed so far can be more easily tweaked now given the rewrite and reuse of code
namely, the following changes can be applied relatively easily:
- whether the dash goes through walls; this is irrespective of object damage, which can still be applied, though it's possible that walls would be unaffected with current values
- whether the dash applies autoattack to encountered mobs - this effect can simply be disabled or replaced for another if it fits the intended gameplay of the ninja better (such as maybe the stun and/or damage from syndicate teleporter's telefrag)
that is to say, ninja dash can now easily go anywhere from old ninja teleport, to what is proposed here but without autoattack, and so on

however, given the requests made of this PR, I'll tick discussion around #11583 as complete and proceed to nerf katana damage as indicated; in exchange, the object damage is buffed so that rwindows can be broken instantly - this still leaves some obstacles such as walls, and partially airlocks or other machines which require at least 3 dashes to break through

slightly more challenging would be the dash recharge/cooldown part - if it has a cooldown, does it still make sense for it to have charges? as it is right now, they recharge pretty quickly
in fact, what I would take a look at is the fact that dash recharges happen in parallel (starting from the moment a charge was used), rather than serially as with the syndicate teleporter (one at a time always) - that is, if you dash thrice in quick succession, 1-2-3, you only have to wait what is the approximate duration of one recharge before being able to use all the dashes again

https://github.com/user-attachments/assets/56f8fcfa-4314-4283-a4a6-4c21b175854e

on one hand this makes the dashes easy to use and even spam when needed, but also feels more powerful than intended

at the very least, what I can do is fix a bug that would prevent the dash from permanently losing charges if the katana was lost mid-dash (``dash_weapon.dm:46``), such as by teleporting on top of an open space and falling down multiz, so that the katana is dropped

**Comment by aramix273** [comment]
time to see if I learnt how to fix merge conflicts

**Comment by PowerfulBacon** [comment]
Looks like it

**Comment by Fronsis** [comment]
> if you dash thrice in quick succession, 1-2-3, you only have to wait what is the approximate duration of one recharge before being able to use all the dashes again

Indeed, that's one of the big issues, their ability to extremely spam it to get away from everyone or just mass-spam it on top of someone to slice them up for a lethal damage without counter play, i'd say they need a more reasonable time so they need to choose wisely and consider the amount of dashes and recharge time before engaging

**Comment by mystery3525** [comment]
I sent you my review changes for better code quality over discord.

File: code/datums/dash_weapon.dm [hunk 1]
```diff
@@ -36,16 +36,30 @@
...
+			if (owner)
+				owner.update_action_buttons_icon()
 			addtimer(CALLBACK(src, PROC_REF(charge)), charge_rate)
 		else
 			to_chat(user, "<span class='warning'>You cannot dash here!</span>")
 
+/datum/action/innate/dash/proc/dashslash(mob/user, turf/slash_location)
+	for(var/mob/living/target in slash_location)//Hit everything in the turf
+		// Skip any mobs that aren't standing, or aren't dense
+		if ((target.body_position == LYING_DOWN) || !target.density || user == target)
```
Comment (PowerfulBacon) [inline code/datums/dash_weapon.dm:58]: If a mob has incorporeal_move at any value then it should not be attacked. incorporeal_move is very poorly coded so isn't checked in a lot of places, but it should be checked here

**Comment by aramix273** [comment]
rather than copy-pasting other charge/cooldown code, I found a way to minimally modify existing code to get the desired effect - testing evidence is included below, and a runtime had to be squashed as to prevent the katana from stopping its recharging forever

https://github.com/user-attachments/assets/3252845e-1efe-48b0-b8df-dfef03c2b8c8

with this done, I will be updating the PR description to include all the changes in both body and player-facing changelog

**Comment by aramix273** [comment]
if the autoattack on dash is not appealing, I can suggest an alternative idea:
xenobiology's stabilized gold has a passive effect that periodically recharges one of the cells in a player's equipment (even those that are installed in devices) by 10%
this effect can inspire an alternative ninja dash effect of stealing charge from the cells of encountered mobs, such as from cyborgs, security cell-powered equipment and so on
discussion can be had if this dash would simply discharge those cells or add the charge to the ninja's suit too, maybe partially (in fact, that effect might require a few more structural changes to make sure the katana knows what suit to add the charge to)

**Comment by Rukofamicom** [comment]
My only outstanding concern about this PR as well as @PowerfulBacon's suggestions is that these will force the ninja to be an aggressive and offensive presence at all times. 

This will leave substantially less room for ninjas to do fun gimmicks and mean that instead of only a fraction of ninjas being unstoppable murder machines, only murderous and far less fun ninjas will remain. A fraction of ninjas will still be unstoppable murder machines, but now closer to 100% of ninjas are trying exclusively to be murderous instead of attempting to be interesting.

**Comment by Rukofamicom** [comment]
I still maintain my prior post as well, regarding this probably being a better implementation of the ability overall and far less frustrating for pursuing security. I'm just apprehensive that this will kill gimmicky ninjas. 

An antagonist can only be gimmicky in the face of valid-thirsty sec if they are as overpowered as ninjas currently are.

**Review by PowerfulBacon — APPROVED** [review]
Seems fine to me

**Comment by Rukofamicom** [comment]
The two specific tests that haven't been performed are not gamebreaking and can be easily tweaked at a later point with a small PR if necessary. I'd like to see how this affects ninja.

**Pull request merged**